### ISSUE TITLE
twisted.web._auth.wrapper causes an "Unhandled error in Deferred"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4411]
|<img alt="honeyman's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| honeyman reported|
|-|-|
|Trac ID|trac#4411|
|Type|defect|
|Created|2010-04-26 13:59:21Z|
|Branch|https://github.com/twisted/twisted/tree/guard-logout-on-error-4411|

Whenever a deferred without any errbacks gets a Failure, the "Unhandled error in Deferred" warning is unconditionally generated. 

Whenever an http.Request.notifyFinish() method is called, a new deferred is generated (rather than, eg., some common one is used). It is a duty of caller to add the proper errback to this deferred, to handle a scenario if the connection is suddenly dropped. If the errback is not added, the "Unhandled error in Deferred" warning is generated.

One of the built-in callers of notifyFinish (and in fact, at the moment, the only one) is ResourceWrapper.render() in HTTPAuthSessionWrapper._loginSucceeded() method in twisted.web._auth.wrapper module - it adds a callback to log out, but doesn't add any errbacks.

<details><summary>Searchable metadata</summary>

```
trac-id__4411 4411
type__defect defect
reporter__honeyman honeyman
priority__normal normal
milestone__ 
branch___branches_guard_logout_on_error_4411 /branches/guard-logout-on-error-4411
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1272290361000000 1272290361000000
changetime__1272640843000000 1272640843000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [28869]) Call logout on notifyFinish errback, too.

refs [#4411](https://github.com/twisted/twisted/issues/4411)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set owner to @exarkun|
|-|-|

1. Move common infrastructure for logout test into shared utility function.

2. That's it I guess, looks good otherwise.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [28881]) Factor common code out of test_logout and test_logoutOnError into a helper method

refs [#4411](https://github.com/twisted/twisted/issues/4411)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [28884]) Merge guard-logout-on-error-4411

Author: exarkun
Reviewer: itamar
Fixes: [#4411](https://github.com/twisted/twisted/issues/4411)

Fix a bug in twisted.web.guard.SessionWrapper which would cause it to fail
to call the realm-provided logout callback if there was an error rendering
the response to the request (eg the client disconnected before it was all
generated).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
